File: src/data/courses/dsa/string-hashing-problems/substring-matching/minimum-window-substring/code/python/minimum_window_substring_sol2.py
class Solution:
    def minWindow(self, s, t):
        from collections import Counter, defaultdict
        need = Counter(t); window = defaultdict(int)
        left = formed = 0; required = len(need)
        min_len, start = float('inf'), 0
        for right in range(len(s)):
            c = s[right]; window[c] += 1
            if c in need and window[c] == need[c]: formed += 1
            while formed == required:
                if right-left+1 < min_len: min_len = right-left+1; start = left
                d = s[left]; window[d] -= 1
                if d in need and window[d] < need[d]: formed -= 1
                left += 1
        return '' if min_len == float('inf') else s[start:start+min_len]

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/minimum-window-substring/code/cpp/minimum_window_substring_sol2.cpp
class Solution {
public:
    // Optimal: O(n) — Expand until valid, shrink to minimize
    string minWindow(string s, string t) {
        unordered_map<char,int> need, window;
        for (char c : t) need[c]++;
        int left = 0, formed = 0, required = need.size();
        int minLen = INT_MAX, start = 0;
        for (int right = 0; right < s.size(); right++) {
            char c = s[right];
            window[c]++;
            if (need.count(c) && window[c] == need[c]) formed++;
            // Shrink while valid
            while (formed == required) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    start = left;
                }
                char d = s[left++];
                window[d]--;
                if (need.count(d) && window[d] < need[d]) formed--;
            }
        }
        return minLen == INT_MAX ? "" : s.substr(start, minLen);
    }
};